File: internal/capture/battery_darwin.c
// apple includes
#include <CoreFoundation/CoreFoundation.h>
#include <IOKit/IOKitLib.h>

// user includes
#include "cf_util.h"
#include "battery_darwin.h"

// @brief   Collects raw battery register values from CoreFoundation
// @return  Returns 0 on success, non-zero if no battery service was found
int collect_battery_raw(battery_raw_t *out) {

    // look up the battery's IOKit service by class name
    io_service_t service = IOServiceGetMatchingService(
        kIOMainPortDefault,
        IOServiceMatching("AppleSmartBattery")
    );

    // pull all properties for this service into one dict.
    CFMutableDictionaryRef props = NULL;
    kern_return_t kr = IORegistryEntryCreateCFProperties(
        service,
        &props,
        kCFAllocatorDefault,
        0
    );
    IOObjectRelease(service);

    // check if properties can be captured from IOKit service
    if (kr != KERN_SUCCESS || props == NULL) {
        return -1;
    }

    // extract each field by key (get_<> calls require address of obj value)
    cf_get_int(props, CFSTR("CurrentCapacity"), &out->current_capacity);
    cf_get_int(props, CFSTR("MaxCapacity"), &out->max_capacity);
    cf_get_int(props, CFSTR("DesignCapacity"), &out->design_capacity);
    cf_get_int(props, CFSTR("NominalChargeCapacity"), &out->nominal_capacity);
    cf_get_int(props, CFSTR("CycleCount"), &out->cycle_count);
    cf_get_int(props, CFSTR("TimeRemaining"), &out->time_remaining);
    cf_get_int(props, CFSTR("AvgTimeToFull"), &out->avg_time_to_full);
    cf_get_int32(props, CFSTR("Amperage"), &out->amperage_ma);
    cf_get_int(props, CFSTR("Voltage"), &out->voltage_mv);
    cf_get_int(props, CFSTR("Temperature"), &out->temperature_centic);
    cf_get_bool(props, CFSTR("IsCharging"), &out->is_charging);
    cf_get_bool(props, CFSTR("ExternalConnected"), &out->is_plugged_in);
    cf_get_bool(props, CFSTR("FullyCharged"), &out->is_fully_charged);

    // free dictionary
    CFRelease(props);

    return 0;
}

File: internal/capture/battery_darwin.go
package capture

// C includes and defs
/*
#cgo LDFLAGS: -framework IOKit -framework CoreFoundation
#include "battery_darwin.h"
*/

import (
	"C"
	"errors"
)

type BatteryRaw struct {
	CurrentCapacity   int
	MaxCapacity       int
	DesignCapacity    int
	NominalCapacity   int
	CycleCount        int
	TimeRemaining     int
	AvgTimeToFull     int
	AmperageMA        int32
	VoltageMV         int
	TemperatureCentiC int
	IsCharging        bool
	IsPluggedIn       bool
	IsFullyCharged    bool
}

func CollectBatteryRaw() (BatteryRaw, error) {
	var raw C.battery_raw_t // pulled from battery_darwin.h

	// capturing battery stats in raw
	if C.collect_battery_raw(&raw) != 0 { // failure
		return BatteryRaw{}, errors.New("no battery present")
	}

	return BatteryRaw{
		CurrentCapacity:   int(raw.current_capacity),
		MaxCapacity:       int(raw.max_capacity),
		DesignCapacity:    int(raw.design_capacity),
		NominalCapacity:   int(raw.nominal_capacity),
		CycleCount:        int(raw.cycle_count),
		TimeRemaining:     int(raw.time_remaining),
		AvgTimeToFull:     int(raw.avg_time_to_full),
		AmperageMA:        int32(raw.amperage_ma),
		VoltageMV:         int(raw.voltage_mv),
		TemperatureCentiC: int(raw.temperature_centic),
		IsCharging:        raw.is_charging != 0,
		IsPluggedIn:       raw.is_plugged_in != 0,
		IsFullyCharged:    raw.is_fully_charged != 0,
	}, nil

}
